languages/java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/SessionFactory.java

```diff
@@ -204,4 +333,50 @@ public SessionFactory build() {
 
     SessionFactory build();
   }
+
+  // Calling it CachedSession but we actually cache the implementing envelope encryption class. Tried
+  // to cache Session instances but the generics and current implementation made it overly difficult to do so.
+  class CachedSession implements EnvelopeEncryption<JSONObject> {
+    private final EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl;
+    // The usageCounter is used to determine if any callers are still using this instance.
+    private final LongAdder usageCounter = new LongAdder();
+    private final String key;
+
+    CachedSession(final EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl, final String key) {
+      this.envelopeEncryptionJsonImpl = envelopeEncryptionJsonImpl;
+      this.key = key;
+    }
+
+    @Override
+    public JSONObject encryptPayload(final byte[] payload) {
+      return envelopeEncryptionJsonImpl.encryptPayload(payload);
+    }
+
+    @Override
+    public byte[] decryptDataRowRecord(final JSONObject dataRowRecord) {
+      return envelopeEncryptionJsonImpl.decryptDataRowRecord(dataRowRecord);
+    }
+
+    @Override
+    public void close() {
+      // Instead of closing the session, we call the release function so it can atomically update the usage counter
+      releaseShared(key);
+    }
+
+    EnvelopeEncryptionJsonImpl getEnvelopeEncryptionJsonImpl() {
+      return envelopeEncryptionJsonImpl;
+    }
+
+    void incrementUsageTracker() {
```

Comment:
The bare external reference counting here seems like a potential place for memory leaks to occur. Maybe something in lang/ref/*Reference may be used instead?

Reply:
Are you referring to this specific part or the caching in general? Been a while since I was working deeply on this, but IIRC when I looked into the soft and weak reference options for the cache, neither were going to feasible for our usage, unfortunately. In general, it is important for users to close the `Session` and `SessionFactory` instances, as indicated in the documentation around usage. Using the cache will make it even more important to close resources properly as it will prevent the finalizer from ever collecting the `Session` objects (finalizer isn't guaranteed to run in the first place).

We can always hop on a Zoom call to dive deeper, if that's easier.